### PR TITLE
db: config: add live_cql_updates_enabled option

### LINKED ISSUE #14355: Add a mechanism to forbid updating LiveUpdatable configuration parameters via CQL
Every configuration parameter (defined in [config.hh](https://github.com/scylladb/scylladb/blob/master/db/config.hh)) has `liveness` value assigned, which controls if a configuration parameter can be changed in runtime or not (in which case it requires a restart). Currently, changing configuration parameters in runtime can be achieved by e.g.:
- modifying `scylla.yaml` and sending `SIGHUP` signal to reload the configuration on the fly,
- updating rows from `system.config` virtual table,
- API call.

A mechanism to forbid updating LiveUpdatable configuration parameters via CQL, i.e. updating `system.config` virtual table, is needed, because we don't want off-premise (cloud) users to reconfigure their nodes themselves (by updating `system.config`) in runtime, but we want to leave that possibility opened for on-premise users. This means that even though a configuration parameter has been defined with `liveness::LiveUpdate`, some users shouldn't be able to change it via CQL in runtime, which is possible now, and for that reason we should treat this issue as a security vulnerability.

Extends https://github.com/scylladb/scylladb/pull/4559

### PR DESCRIPTION
**Pull request by ptrsmrn**
If `live_cql_updates_enabled` is set to true, configuration parameters defined with `liveness::LiveUpdate` option can be updated in the runtime with CQL, i.e. by updating `system.config` virtual table.
If we don't want any configuration parameter to be changed in the
runtime by updating `system.config` virtual table, this option should be
set to false. This option should be set to false for e.g. cloud users,
who can only perform CQL queries, and should not be able to change
scylla's configuration on the fly.

Current implemenatation is generic, but has a small drawback - messages
returned to the user can be not fully accurate, consider:
```
cqlsh> UPDATE system.config SET value='2' WHERE name='task_ttl_in_seconds';
WriteFailure: Error from server: code=1500 [Replica(s) failed to execute write] message="option is not live-updateable" info={'failures': 1, 'received_responses': 0, 'required_responses': 1, 'consistency': 'ONE'}
```
where `task_ttl_in_seconds` has been defined with
`liveness::LiveUpdate`, but because `live_cql_updates_enabled` is set to
`false` in `scylla.yaml,` `task_ttl_in_seconds` cannot be modified in the
runtime by updating `system.config` virtual table.

Fixes https://github.com/scylladb/scylladb/issues/14355

### CODE REVIEW

**Comment by ptrsmrn** [comment]
TESTING:

- CQL updates are blocked, for the message please see the description
- sending `SIGHUP` does not reload the config if `live_updates_enabled = false`
- calling API:
```
(with live_updates_enabled = true. API call returns the old config value)
[piotrs@new-host-3 scylla2]$ curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' 'http://localhost:10000/task_manager/ttl?ttl=3'
2[piotrs@new-host-3 scylla2]$ curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' 'http://localhost:10000/task_manager/ttl?ttl=4'
3[piotrs@new-host-3 scylla2]$

(with live_updates_enabled = false)
[piotrs@new-host-3 scylla2]$ curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' 'http://localhost:10000/task_manager/ttl?ttl=4'
10[piotrs@new-host-3 scylla2]$ curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' 'http://localhost:10000/task_manager/ttl?ttl=4'
10[piotrs@new-host-3 scylla2]$ curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' 'http://localhost:10000/task_manager/ttl?ttl=4'
10
```
I also intend to update `TestUpdateableConfig` suite in dtests.

**Comment by nyh** [comment]
> TESTING:
> 
>     * CQL updates are blocked, for the example please see the description
> 
>     * sending `SIGHUP` does reload a config parameter if `live_updates_enabled = false` - changes are visible in system.config table

I guess because of this you renamed "live_updates_enabled" in this comment to "live_cql_updates_enabled" which is now described in the cover letter.

So I think the name of this configuration option could be even clearer as "writeable_system_config" or "cql_writeable_system_config".  Basically, this option has nothing to do with "LiveUpdate", not to mention that the phrase "live update" would be a big mystery to users ("ive update of what? why isn't the word "configuration" somehow mentioned in this name?). The option just makes the system.config CQL table writeable, or read-only, so maybe the option name should say that.

>     * calling API also works (sets the new value and returns the old one) - changes are visible in system.config table

Are we fine with the REST API being able to change configuration? Why? Maybe we need another option to disable that?

**Comment by nyh** [comment]
> ```
> cqlsh> UPDATE system.config SET value='2' WHERE name='task_ttl_in_seconds';
> WriteFailure: Error from server: code=1500 [Replica(s) failed to execute write] message="option is not live-updateable" info={'failures': 1, 'received_responses': 0, 'required_responses': 1, 'consistency': 'ONE'}
> ```

A thought: could we make the error the same error you get when you are **unauthorized** (by roles, etc.)? I.e., you can't write to the `system.config` table because, well, you're not allowed to write to it. Ideally, the error shouldn't look like an ugly replica-side error like you quited above (and definitely not have misleading text about option not being live-updatable) - but rather should be a clean coordinator-side authorization error.

File: conf/scylla.yaml [hunk 1]
```diff
@@ -590,3 +590,10 @@ consistent_cluster_management: true
 # New clusters have this option set to `true` by scylla.yaml (which overrides the default `warn`)
 # to make sure that trying to create an invalid view causes an error.
 strict_is_not_null_in_views: true
+
+# If set to true, configuration parameters defined with LiveUpdate option can be updated in the runtime.
+# If we don't want any configuration parameter to be changed in the runtime by any means, i.e.
+# reloading scylla.yaml with SIGHUP, updating system.config virtual table or calling API,
+# this option should be set to false. This option should be set to false for e.g. cloud users, for whom
+# scylla's configuration should be changed only by support engineers.
+live_updates_enabled: true
```
Comment (ptrsmrn) [inline conf/scylla.yaml:599]: @tzach should we, by default, allow users to update configuration parameters in CQL?

**Comment by ptrsmrn** [comment]
> > TESTING:
> > ```
> > * CQL updates are blocked, for the example please see the description
> > 
> > * sending `SIGHUP` does reload a config parameter if `live_updates_enabled = false` - changes are visible in system.config table
> > ```
> 
> I guess because of this you renamed "live_updates_enabled" in this comment to "live_cql_updates_enabled" which is now described in the cover letter.

Initially this config option I introduce was supposed to block updating configuration parameters in runtime by any mechanism (CQL, SIGHUP, API), but later we agreed it's enough to only block CQL live-updates, and so I reworked the PR... and pushed only new commit message - sorry. Now the code changes are also up-to-date.
> 
> So I think the name of this configuration option could be even clearer as "writeable_system_config" or "cql_writeable_system_config". Basically, this option has nothing to do with "LiveUpdate", not to mention that the phrase "live update" would be a big mystery to users ("ive update of what? why isn't the word "configuration" somehow mentioned in this name?). The option just makes the system.config CQL table writeable, or read-only, so maybe the option name should say that.

I agree that the current name can be improved. Perpahs "cql_writeable_system_config" or "make_system_config_read_only" would be good, because only "writeable_system_config" wouldn't tell the whole story, as system.config would remain writeable, just not by CQL.
> 
> > ```
> > * calling API also works (sets the new value and returns the old one) - changes are visible in system.config table
> > ```
> 
> Are we fine with the REST API being able to change configuration? Why? Maybe we need another option to disable that?

That's fine. The whole point here is to address the vulnerability where cloud users can modify configuration themselves.  Leaving API/SIGHUP working is good for our support engineers to be able to use them to update nodes without any downtime.

**Comment by ptrsmrn** [comment]
> > ```
> > cqlsh> UPDATE system.config SET value='2' WHERE name='task_ttl_in_seconds';
> > WriteFailure: Error from server: code=1500 [Replica(s) failed to execute write] message="option is not live-updateable" info={'failures': 1, 'received_responses': 0, 'required_responses': 1, 'consistency': 'ONE'}
> > ```
> 
> A thought: could we make the error the same error you get when you are **unauthorized** (by roles, etc.)? I.e., you can't write to the `system.config` table because, well, you're not allowed to write to it. Ideally, the error shouldn't look like an ugly replica-side error like you quited above (and definitely not have misleading text about option not being live-updatable) - but rather should be a clean coordinator-side authorization error.

It's definitely doable, will just require adjusting error messages in different code places. Keeping the implementation generic, as it currently is, limits lines of code changed.

File: conf/scylla.yaml (hunk 1 above)
Comment (eliransin) [reply conf/scylla.yaml]: @tzach ping

Comment (ptrsmrn) [reply conf/scylla.yaml]: @tzach agreed with my comment to let on-premise (oss & enterprise) users change config params via CQL, so we can leave this param set to `true` by default, and it'll have to be set to `false` for cloud deployments.

**Review by eliransin — APPROVED** [review]

**Comment by eliransin** [comment]
@ptrsmrn we just need to figure out why CI failed.

**Comment by avikivity** [comment]
How are these configuration changes persisted?

File: db/config.cc [hunk 2]
```diff
@@ -997,13 +997,14 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , relabel_config_file(this, "relabel_config_file", value_status::Used, "", "Optionally, read relabel config from file")
     , object_storage_config_file(this, "object_storage_config_file", value_status::Used, "", "Optionally, read object-storage endpoints config from file")
     , minimum_keyspace_rf(this, "minimum_keyspace_rf", liveness::LiveUpdate, value_status::Used, 0, "The minimum allowed replication factor when creating or altering a keyspace.")
-    , auth_superuser_name(this, "auth_superuser_name", value_status::Used, "", 
+    , auth_superuser_name(this, "auth_superuser_name", value_status::Used, "",
```
Comment (avikivity) [inline db/config.cc:1000]: Why this change?

File: db/config.cc [hunk 3]
```diff
@@ -997,13 +997,14 @@ db::config::config(std::shared_ptr<db::extensions> exts)
...
         "Initial authentication super username. Ignored if authentication tables already contain a super user")
     , auth_superuser_salted_password(this, "auth_superuser_salted_password", value_status::Used, "", 
         "Initial authentication super user salted password. Create using mkpassword or similar. The hashing algorithm used must be available on the node host. "
         "Ignored if authentication tables already contain a super user password.")
     , auth_certificate_role_queries(this, "auth_certificate_role_queries", value_status::Used, { { { "source", "SUBJECT" }, {"query", "CN=([^,]+)" } } },
         "Regular expression used by CertificateAuthenticator to extract role name from an accepted transport authentication certificate subject info.")
+    , live_cql_updates_enabled(this, "live_cql_updates_enabled", liveness::MustRestart, value_status::Used, true, "If set to true, configuration parameters defined with LiveUpdate can be updated in runtime via CQL (by updating system.config virtual table), otherwise they can't.")
```
Comment (avikivity) [inline db/config.cc:1007]: Why MustRestart? It's hostile to operators.

Comment (tzach) [reply db/config.cc:1007]: The motivation is allowing the DBA to block random users from updating the config for everyone.
These config updates might be dangerous, inconsistent (between nodes), and not persistent.

Another alternative is using RBAC.

Comment (avikivity) [reply db/config.cc:1007]: It's completely unrelated.

Comment (tzach) [reply db/config.cc:1007]: How so? What stops a user from enabling live_cql_updates_enabled and then updating all other config parameters? can we at least limit live_cql_updates_enabled to REST API update?

Comment (avikivity) [reply db/config.cc:1007]: LiveUpdate means you can change value in scylla.yaml and send SIGHUP instead of a node restart. If the user doesn't have access to scylla.yaml, they can't change anything.

Comment (ptrsmrn) [reply db/config.cc:1007]: Correct, and the main motivation was for cloud not to be able to change config options via system.config table in runtime, but still leave this door opened for on-premise. That's why this param is enabled by default, and will have to be disabled for cloud (set to false in scylla.yaml, once this change is deployed). I assume cloud cannot send signals to reload configuration.

**Comment by ptrsmrn** [comment]
> How are these configuration changes persisted?

@avikivity sorry, I am not sure what you mean, can you please elaborate? 
If someone changes config param on the fly, because live_cql_updates_enabled is enabled, then how such a change is persisted is outside of the scope of this PR, because this PR merely limits ability of users to perform such a change.
Other than that, live_cql_updates_enabled can be persisted in scylla.yaml.

File: conf/scylla.yaml [hunk 4]
```diff
@@ -590,3 +590,10 @@ consistent_cluster_management: true
 # New clusters have this option set to `true` by scylla.yaml (which overrides the default `warn`)
 # to make sure that trying to create an invalid view causes an error.
 strict_is_not_null_in_views: true
+
+# If set to true, configuration parameters defined with LiveUpdate option can be updated in runtime with CQL
+# by updating system.config virtual table. If we don't want any configuration parameter to be changed in runtime
+# via CQL, this option should be set to false. This parameter doesn't impose any limits on other mechanisms updating
+# configuration parameters in runtime, e.g. sending SIGHUP or using API. This option should be set to false
+# e.g. for cloud users, for whom scylla's configuration should be changed only by support engineers.
+# live_cql_updates_enabled: true
```
Comment (ptrsmrn) [inline conf/scylla.yaml:599]: Param name should be more explicit, think of something like: live_updateable_config_options_changeable_in_runtime

File: db/config.cc (hunk 2 above)
Comment (avikivity) [reply db/config.cc:1000]: Yes, we generally try to limit patches to their scope. One reason is that it reduces effort when backporting (the unrelated change might need extra treatment), another is to preserve `git log` and `git blame` cleanliness, but the real reason is that that's how we did it when working on the kernel.

File: db/config.cc (hunk 3 above)
Comment (avikivity) [reply db/config.cc:1007]: The cloud can send signals to reload configuration. Either when an operator does this manually, or one day when they automate it automatically. In any case we need to try to make the system good and free of rough edges (rolling restart = rough edge) and not tailor it to a specific user.

**Comment by avikivity** [comment]
> > How are these configuration changes persisted?
> 
> @avikivity sorry, I am not sure what you mean, can you please elaborate? If someone changes config param on the fly, because live_cql_updates_enabled is enabled, then how such a change is persisted is outside of the scope of this PR, because this PR merely limits ability of users to perform such a change. Other than that, live_cql_updates_enabled can be persisted in scylla.yaml.

Agree it's out of scope for this series. But it needs to be addressed, or the value from these patches is diminished.

There's also a problem in that the changes are applied per-node rather than for the entire cluster. For a large cluster, this would mean connecting to each node individually, and hoping cqlsh remains on the same node.

File: conf/scylla.yaml (hunk 4 above)
Comment (ptrsmrn) [reply conf/scylla.yaml]: Changed to `live_updatable_config_params_changeable_via_cql`.

**Comment by ptrsmrn** [comment]
@avikivity ping

**Comment by avikivity** [comment]
Please use Fixes, not Resolves, in the commit changelog so our machinery can pick it up.

**Comment by ptrsmrn** [comment]
> Please use Fixes, not Resolves, in the commit changelog so our machinery can pick it up.

OK, done.

**Comment by ptrsmrn** [comment]
@nyh can you help me understand why this PR got close without actually merging the changes? Was it opened for too long without anybody approving?
I'd like to have it merged, eventually.

**Comment by fruch** [comment]
> @nyh can you help me understand why this PR got close without actually merging the changes? Was it opened for too long without anybody approving?
> I'd like to have it merged, eventually.

Promoter closed it, cause this commit reached the master branch:
https://github.com/scylladb/scylladb/commit/34c3688017e9acee2aff58d47e43719863c08c15

You are saying it's not there anymore?

**Comment by ptrsmrn** [comment]
> > @nyh can you help me understand why this PR got close without actually merging the changes? Was it opened for too long without anybody approving?
> > I'd like to have it merged, eventually.
> 
> Promoter closed it, cause this commit reached the master branch: [34c3688](https://github.com/scylladb/scylladb/commit/34c3688017e9acee2aff58d47e43719863c08c15)
> 
> You are saying it's not there anymore?

Thanks, I got confused by github's message (pic attached).
image
